Use FakeTensor cache for subclass inner tensors

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132414
* #130292
* __->__ #131803
* #131916

Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with @bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

soulitzer committed:
Use FakeTensor cache for subclass inner tensors

[ghstack-poisoned]

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"




[ghstack-poisoned]

YuqingJ commented:
@yuqingj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60406773).

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -809,13 +809,16 @@ def _empty_create_subclass(
                 # We are hitting plain meta_desc tensor so actually
```
I have an alternate proposal: just make this function call recursively back to `meta_tensor`, which as you say, handles caching. Does that work?

jbschlosser replied:
agreed with this! this feels like the right thing to do and I'm curious if there's some obscure reason it wasn't

ezyang replied:
I'm pretty sure there's no good reason, there are a bunch of places where we don't recurse back to top level when we should. It's been on my TODO list to force all the recurse-y calls to actually recurse, but I never got around to it.

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"


Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

[ghstack-poisoned]

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"


Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

[ghstack-poisoned]

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"


Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

[ghstack-poisoned]

soulitzer commented on test/functorch/test_aotdispatch.py:
```diff
@@ -841,8 +841,10 @@ def f(x):
+        a4 = a.clone().detach().requires_grad_()
         aa = TwoTensor(a, a2)
-        aa2 = aa.clone().detach().requires_grad_()
+        aa2 = TwoTensor(a3, a4)
```
One difference between calling recursively back to `meta_tensor` and using `callback` is that using `meta_tensor` we now faithfully replicate the requires_grad of original tensor.

With this change, this test fails during fakification when constructing the outer TwoTensor out of the two Tensors because the requires_grad mismatch. 

TwoTensor's requires_grad is determined by the requires_grad of the inner tensor, which for aa2 have actually been detached. 

The reason this succeeds in eager because we manually set aa2's requires_grad to True.

bdhirsh replied:
tbh I think we should fix this on the TwoTensor side - it's a bit of an anti-pattern that I originally wrote `TwoTensor` to "inherit" its requires_grad from the inner tensor. We should just:

(1) not set requires_grad in anyway within the `TwoTensor()` constructor

(2) add a differentiable constructor for TwoTensor that wraps the `TwoTensor()` constructor into an autograd.Function`. e.g. `TwoTensor.from_tensors(a, b)`

does that sound good to you? (it's also more inline with what we do for float8/dtensor)

ezyang replied:
Yeah, inheriting requires grad from inner is just wrong, don't do that lol

soulitzer replied:
> not set requires_grad in anyway within the TwoTensor() constructor

Sounds good! There seem to be some places in the tests to update however, so I'm going to do it in a follow up to keep this PR small.

ezyang reviewed (approved):
music to my ears

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"


Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

[ghstack-poisoned]

soulitzer committed:
Update on "Use FakeTensor cache for subclass inner tensors"


Rewrite of original PR in https://github.com/pytorch/pytorch/pull/130291 

To answer review comments from https://github.com/pytorch/pytorch/pull/130291#pullrequestreview-2166671953:

> At a higher level, do we need this? 

Today, this should not change the behavior of anything. But an invariant of "same tensor always corresponds to the same FakeTensor" is nice (from discussion with bdhirsh).

> Why does this happen?

Today, both dynamo and meta_utils do some recursion when it comes to FakeTensors. So whenever we fakify a subclass, the process would roughly like:

```
wrap_to_fake (subclass)
   meta_utils (subclass)
      meta_utils (values) -> not cached because we use callback
      meta_utils(offsets) -> not cached because we use callback
  wrap_to_fake (values)
  wrap_to_fake (offsets) -> cached because we rely on top-level meta_utils
```

However, we know that:
- Caching only occurs at the top-level of meta_utils. 
- The return value of the top-level wrap_to_fake is returned.

This means that after all of this:
- The fakified subclass holds inner FakeTensors that are NOT part of the cache
- values/offsets are Fakified a second time, and those instances are cached.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406773](https://our.internmc.facebook.com/intern/diff/D60406773)

[ghstack-poisoned]
